src/lib/Sympa/Request/Handler/add_list_admin.pm

```diff
@@ -0,0 +1,149 @@
+# -*- indent-tabs-mode: nil; -*-
+# vim:ft=perl:et:sw=4
+# $Id$
+
+# Sympa - SYsteme de Multi-Postage Automatique
+#
+# Copyright 2017 The Sympa Community. See the AUTHORS.md file at the top-level
+# directory of this distribution and at
+# <https://github.com/sympa-community/sympa.git>.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package Sympa::Request::Handler::add_list_admin;
+
+use strict;
+use warnings;
+use Data::Dumper;
+
+use Sympa;
+use Sympa::List;
+use Sympa::Log;
+use Sympa::Robot;
+use Sympa::Tools::Text;
+
+use base qw(Sympa::Request::Handler);
+
+my $log = Sympa::Log->instance;
+
+use constant _action_scenario => undef; # Only privileged owners and listmasters allowed.
+use constant _context_class   => 'Sympa::List';
+
+sub _twist {
+    my $self    = shift;
+    my $request = shift;
+
+    my $list    = $request->{context};
+    my $listname   = $list->{'name'};
+    my $robot   = $list->{'domain'};
+    my $role = $request->{role};
+    my $sender = $request->{sender};
+    my $user;
+    $user->{email} = Sympa::Tools::Text::canonic_email($request->{email});
+    $user->{visibility} = $request->{visibility};
+    $user->{profile} = $request->{profile};
+    $user->{reception} = $request->{reception};
+    $user->{gecos} = $request->{gecos};
+    $user->{info} = $request->{info};
+
+    unless ($user->{email} and $robot and $role) {
+        die "Missing incoming parameter. robot: $robot, role: $role, user: ".$user->{email};
+    }
+    # Check if user is already admin of the list.
+    if ($list->is_admin($role, $user->{email})) {
+        unless ($list->update_list_admin($user->{email}, $role, $user)) {
```

Comment:
I agree.
Adding is not the same operation as modify, so I'll fix my code to prevent modification in this handler.
modify should be in another handler.

Reply:
(Even though, for example in the web interface, if the "add" handler fails, we could try the "update" after, to prevent bad user experience).

Reply:
Just so you know I AM modifying the code right now. But I have a problem with list admin cache. I'm reading the code to fully understand my problem and will carry on once I have solved my issue.

Reply:
Fixed in commit 105cf45b5185420e08bf327cc9fc646fd6f9240c.
See also PR #583 for a change I had to do to make my tests to succeed.

Reply:
I'm OK.

Reply:
Cool. I'm resolving this conversation.